File: src/Photokompressor/Core/Presets.cs
namespace Photokompressor.Core;

public static class Presets
{
    public record JpegParams(int Q, bool ForceSubsampleOn);
    public record PngParams(bool Palette, int Q, double Dither, int Effort);
    public record WebpParams(int Q, int Effort, int AlphaQ);

    public static JpegParams Jpeg(QualityPreset preset) => preset switch
    {
        QualityPreset.High => new JpegParams(82, ForceSubsampleOn: false),
        QualityPreset.Balanced => new JpegParams(72, ForceSubsampleOn: true),
        QualityPreset.Smallest => new JpegParams(58, ForceSubsampleOn: true),
        _ => new JpegParams(72, true),
    };

    public static PngParams Png(QualityPreset preset) => preset switch
    {
        QualityPreset.High => new PngParams(Palette: false, Q: 100, Dither: 1.0, Effort: 7),
        QualityPreset.Balanced => new PngParams(Palette: true, Q: 90, Dither: 1.0, Effort: 7),
        QualityPreset.Smallest => new PngParams(Palette: true, Q: 65, Dither: 0.8, Effort: 10),
        _ => new PngParams(true, 90, 1.0, 7),
    };

    public static WebpParams Webp(QualityPreset preset) => preset switch
    {
        QualityPreset.High => new WebpParams(80, Effort: 4, AlphaQ: 90),
        QualityPreset.Balanced => new WebpParams(68, Effort: 5, AlphaQ: 85),
        QualityPreset.Smallest => new WebpParams(52, Effort: 6, AlphaQ: 80),
        _ => new WebpParams(68, 5, 85),
    };
}

File: macos/Sources/PhotokompressorCore/Presets.swift
import Foundation

/// Numeric tuning per preset — kept identical to the Windows build's Presets.cs
/// so output size/quality tradeoffs match across platforms.
public enum Presets {
    public struct JpegParams { public let q: Int32; public let forceSubsampleOn: Bool }
    public struct PngParams { public let palette: Bool; public let q: Int32; public let dither: Double; public let effort: Int32 }
    public struct WebpParams { public let q: Int32; public let effort: Int32; public let alphaQ: Int32 }

    public static func jpeg(_ preset: QualityPreset) -> JpegParams {
        switch preset {
        case .high: return JpegParams(q: 82, forceSubsampleOn: false)
        case .balanced: return JpegParams(q: 72, forceSubsampleOn: true)
        case .smallest: return JpegParams(q: 58, forceSubsampleOn: true)
        }
    }

    public static func png(_ preset: QualityPreset) -> PngParams {
        switch preset {
        case .high: return PngParams(palette: false, q: 100, dither: 1.0, effort: 7)
        case .balanced: return PngParams(palette: true, q: 90, dither: 1.0, effort: 7)
        case .smallest: return PngParams(palette: true, q: 65, dither: 0.8, effort: 10)
        }
    }

    public static func webp(_ preset: QualityPreset) -> WebpParams {
        switch preset {
        case .high: return WebpParams(q: 80, effort: 4, alphaQ: 90)
        case .balanced: return WebpParams(q: 68, effort: 5, alphaQ: 85)
        case .smallest: return WebpParams(q: 52, effort: 6, alphaQ: 80)
        }
    }
}
